src/lib/coda_base/protocol_constants_checked.ml

```diff
@@ -0,0 +1,121 @@
+[%%import
+"/src/config.mlh"]
+
+open Core_kernel
+
+[%%ifdef
+consensus_mechanism]
+
+open Snark_params.Tick
+
+[%%else]
+
+open Snark_params_nonconsensus
+module Random_oracle = Random_oracle_nonconsensus.Random_oracle
+
+[%%endif]
+
+module T = Coda_numbers.Length
+
+(*constants actually required for blockchain snark*)
+(* k
+  ,c
+  ,slots_per_epoch
+  ,slots_per_sub_window
+  ,sub_windows_per_window
+  ,checkpoint_window_size_in_slots
+  ,block_window_duration_ms*)
+
+module Poly = Genesis_constants.Protocol.Poly
+
+module Value = struct
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t =
+        (T.Stable.V1.t, T.Stable.V1.t, Block_time.Stable.V1.t) Poly.Stable.V1.t
+      [@@deriving eq, ord, hash, sexp, to_yojson, compare]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  type t = Stable.Latest.t [@@deriving to_yojson, eq, sexp, compare]
+
+  let gen : t Quickcheck.Generator.t =
+    let open Quickcheck.Let_syntax in
+    let%bind k = Int.gen_incl 1 5000 in
+    let%bind delta = Int.gen_incl 0 5000 in
+    (*TODO: Bug -> Block_time.(to_time x |> of_time) != x for certain values.
```

Comment:
Yes

> LGTM. We should land this as is, but I have an idea for a different way to organize the runtime constants which would have less impact on the code.
> 
> While the constants need to be loaded at the beginning of the program's runtime now, they are still constant. Rather than passing a value containing all the constants between all the functions which refer to them, we could instead have a single module which contains `val constants : t option ref` which can only be set once, along with some shorthand getters. All the code which refers to constant values can just assume the constants will have been loaded before that code is execute, which should be a safe assumption since the only portion of the code which this would ever not be true for would be the CLI entrypoint.

Not quite. You'd have to set it for runtime_genesis_ledger and other exes that run any part of the code that uses  these constants. I started with this design but decided against it because of this dependency. It also maintains some level of consistency between in-snark and out-of-snark functions because the constants gets passed to the functions.

Reply:
Ok. We should open an issue to investigate this and potentially swap out `Core.Time` for `Core.Time_ns` in most places.